src/eduid_userdb/credentials/password.py

```diff
@@ -32,134 +32,57 @@
-        raise_on_unknown: bool = True,
-        called_directly: bool = True,
-    ):
-        data_in = data
-        data = copy.copy(data_in)  # to not modify callers data
-
-        if data is None:
-            if created_ts is None:
-                created_ts = True
-            data = dict(
-                id=credential_id, salt=salt, is_generated=is_generated, created_by=application, created_ts=created_ts,
-            )
-        elif 'created_ts' not in data:
-            data['created_ts'] = True
-
-        if 'source' in data:  # TODO: Load and save all users in the database to replace source with created_by
-            data['created_by'] = data.pop('source')
-        super().__init__(data, called_directly=called_directly)
-        if 'id' in data:  # TODO: Load and save all users in the database to replace id with credential_id
-            data['credential_id'] = data.pop('id')
-        self.is_generated = data.pop('is_generated', False)
-        self.credential_id = data.pop('credential_id')
-        self.salt = data.pop('salt')
-
-        leftovers = data.keys()
-        if leftovers:
-            if raise_on_unknown:
-                raise UserHasUnknownData('Password {!r} unknown data: {!r}'.format(self.key, leftovers))
-            # Just keep everything that is left as-is
-            self._data.update(data)
-
-    @classmethod
-    def from_dict(cls: Type['Password'], data: Dict[str, Any], raise_on_unknown: bool = True) -> 'Password':
-        """
-        Construct password credential from a data dict.
-        """
-        return cls(data=data, raise_on_unknown=raise_on_unknown, called_directly=False)
+@dataclass
+class _PasswordRequired:
+    """
+    Required fields for Password
+    """
 
-    @property
-    def key(self) -> str:
-        """
-        Return the element that is used as key.
-        """
-        return self.credential_id
+    credential_id: str
+    salt: str
 
-    @property
-    def credential_id(self) -> str:
-        """
-        This is a reference to the ObjectId in the authentication private database.
-        """
-        return self._data['credential_id']
+    def __post_init__(self):
+        # backwards compat
+        if isinstance(self.credential_id, bson.ObjectId):
+            self.credential_id = str(self.credential_id)
 
-    @credential_id.setter
-    def credential_id(self, value: Union[ObjectId, str]):
-        """
-        :param value: Reference to the password credential in the authn backend db.
-        """
-        if isinstance(value, ObjectId):
-            # backwards compatibility
-            value = str(value)
-        if not isinstance(value, str):
-            raise UserDBValueError("Invalid 'credential_id': {!r}".format(value))
-        self._data['credential_id'] = value
 
-    @property
-    def salt(self) -> str:
-        """
-        Password salt.
-        """
-        return self._data['salt']
+@dataclass(eq=False, unsafe_hash=True)
```

Comment:
Same here as for Fido and Webauthn. credential_id can probably be used for the `__hash__` method.